### PR TITLE
[diffusion] support resident layers for DiT

### PR DESCRIPTION
**Pull request by AgainstEntropy**
## Motivation

`--dit-layerwise-offload` is useful on GPUs with limited VRAM. However, all layers are streamed H2D at each denoising step. And tuning `--dit-offload-prefetch-size` doesn't speed up as VRAM increases (see table below).

A new flag `--dit-layerwise-resident-layers` is introduced here to keep leading layers resident across denoising steps, so that only tail layers needs streaming. This flag is intended to use together with `--dit-layerwise-offload`.

## Accuracy Tests

Verified offline that with different `--dit-layerwise-resident-layers` it produces byte-identical output images.

## Speed Tests and Profiling

Tested on 1× RTX 5090, 32 GB with

- `Qwen/Qwen-Image` (60 layers)

| config | peak VRAM (GB) | denoise (s) | OOM |
|---|---|---|---|
| `--dit-layerwise-offload` prefetch 1 | 5.43 | 56.60 | — |
| `--dit-layerwise-offload` prefetch 16 | **24.47** | 56.12 | — |
| `--dit-layerwise-offload` prefetch 20 | — | — | OOM @ denoise step 0 |
| | | | |
| `--dit-layerwise-resident-layers 30` prefetch 1 | **24.47** | **36.68** | — |
| `--dit-layerwise-resident-layers 34` prefetch 1 | 25.11 | **33.97** | — |
| `--dit-layerwise-resident-layers 38` prefetch 1 | — | — | OOM @ denoise step 0 |

- `FLUX.1-dev` (dual-stack, 19 + 38 layers)

| config | peak VRAM (GB) | denoise (s) |
|---|---|---|
| `--dit-layerwise-offload` prefetch 1 | 6.56 | 16.96 |
| `--dit-layerwise-offload` prefetch 4 | 12.09 | 16.95 |
| `--dit-layerwise-offload` prefetch 6 | 15.78 | 16.70 |
| | | |
| `--dit-layerwise-resident-layers 0.5` prefetch 1 | **14.56** | **13.06** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30410965710](https://github.com/sgl-project/sglang/actions/runs/30410965710)
Latest PR Test (Extra): :x: [Run #30410965629](https://github.com/sgl-project/sglang/actions/runs/30410965629)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py [hunk 1]
```diff
@@ -42,12 +42,20 @@ def __init__(
         enabled: bool,
         pin_cpu_memory: bool = True,
         prefetch_size: int = 1,
+        resident_layers: int = 0,
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py:45]: for better latency, maybe keep this value same as before?

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py [hunk 2]
```diff
@@ -254,11 +262,42 @@ def prepare_for_next_req(self, non_blocking=True):
...
+        if self._residency_initialized:
+            return
+        self._residency_initialized = True
+        if self._configured_resident_layers <= 0:
+            return
+        self._num_resident_layers = min(
+            self.num_layers, self._configured_resident_layers
+        )
+
+        logger.info(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py]: could we merge and simplify the initialization log

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py [hunk 3]
```diff
@@ -674,6 +728,15 @@ def is_layerwise_offloaded_module(module: torch.nn.Module) -> bool:
     )
 
 
+def is_resident_layerwise_module(module: torch.nn.Module) -> bool:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py:720]: `@lru_cache`?

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py (hunk 1 above)
Comment (AgainstEntropy) [reply python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py:45]: Do you mean `resident_layers`? This is a new value, and 0 by default gives the same behavior as before

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py (hunk 3 above)
Comment (AgainstEntropy) [reply python/sglang/multimodal_gen/runtime/managers/memory_managers/layerwise_offload.py:720]: `is_resident_layerwise_module` is cheap `isinstance` check, and it's not on the hot path of denoising loop.

I would prefer keeping no `@lru_cache` as how `is_layerwise_offloaded_module` does now.

**Review by mickqian — APPROVED** [review]
